### ISSUE TITLE
Mac OS X MariaDB 10.3.13 binaries status unclear - working or not? If NOK, how to fix?

### ISSUE THREAD

**Issue by vorburger** [vorburger/MariaDB4j#288]
So in #235 for #230, @srbala contributed MariaDB 10.3.13 binaries (before we had 10.2.11). For Linux, the new 10.3.13 binaries clearly work (given that the Travis CI for #235 passed!), for Windows I'm guessing it also does (I've not personally tested it; no Windows here).

For Mac the status isn't super clear, to me... @jtmuhone, of #229 fame, during review of #235 provided some feedback which makes me think that it as-is currently (just after merging #229) it may not just work out of the box on Mac? I've not personally tested it (no Mac)... would someone else like to test this, and provide some feedback here? Or perhaps @srbala and/or @jtmuhone you can chime in here? (Again, I'm so sorry for the delay in processing this.)

**Comment by vorburger** [comment]
The Mac related problems actually aren't new to the MariaDB binary 10.3.13 (or 10.3.17 coming up in #292), issue #278 illustrates a problem a Mac user had even before #235 already..

**Comment by martin-traverse** [comment]
sudo mkdir -p /usr/local/opt/openssl/lib
sudo ln -s /usr/lib/libssl.dylib /usr/local/opt/openssl/lib/libssl.1.0.0.dylib
sudo ln -s /usr/lib/libcrypto.dylib opt/openssl/lib/libcrypto.1.0.0.dylib

It's icky but it works for me on Mojave and avoids installing Homebrew. Suspect it won't work in Catalina due to the read-only root filesystem (and rightly so)! For my project, we'll probably disable local tests on macOS for the MySQL dialect until this fixed. Or until someone really, really wants to test against MySQL on a shiny new MacBook ;-)

**Comment by venkyvb** [comment]
@martin-traverse Just wanted to check if you tried your hack in Catalina? I am running into the same issue and currently the only option seems to be to downgrade openssl to 1.0.2t which breaks a bunch of other things.

**Comment by martin-traverse** [comment]
Yes I did, and to my surprise it worked!

**Comment by slangeberg** [comment]
Manual workaround - upgrade to mariadb locally that has openssl 1.1 support

https://formulae.brew.sh/formula/mariadb@10.1

Notice - Depends On: 
   openssl@1.1 | 1.1.1i | Cryptography and SSL/TLS Toolkit

`brew install mariadb@10.1`

Use spring config to point to local maria db
```
# Bypass issues with bundled version depends on openssl 1.0
mariaDB4j.baseDir=/usr/local/Cellar/mariadb@10.1/10.1.48
#mariaDB4j.libDir=
mariaDB4j.unpack=false
```
Integration tests seem to work fine

**Comment by A248** [comment]
Am I missing something, or is this the same issue as #48 ? Both of these issues are resolved by symlinking the openssl binaries.

The reason this issue, 288, was ever raised is because a user received an issue with the PR: https://github.com/vorburger/MariaDB4j/pull/235#issuecomment-474375215 . However, the error reported seems to be nearly the same error as with #48 . And @martin-traverse has described the solution to this issue, which is the same solution to #48 , assuming the omission of /usr/local/ is insignificant.

Past MariaDB4j versions have been released in spite of this issue. I am on a Mac myself (unfortunately), and while the situation is not perfect, it is fixable with a couple of symlinks.

**Comment by slangeberg** [comment]
@A248 the symlink fix is global right? So if i have different tools which require different versions of openssl then i can't use both tools at the same time. Right?

**Comment by A248** [comment]
@slangeberg I'm not really sure what will happen. There are several things to note.

Users in #48 report that this problem occurs because Mac uses openssl 1.1.0. The solution is to symlink the 1.1.x version to the 1.0.x version, that way MariaDB4j can use the file named as if it was a 1.0.x binary even though it is really a 1.1.x binary. That has apparently worked for many people and it also worked for me. I suspected this was due to semantic versioning - 1.1.x should be compatible with 1.0.x, right?

Well, it turns out openssl does not follow semantic versioning, and 1.1.x is not strictly compatible with 1.0.x. ([ref1](https://wiki.openssl.org/index.php/Versioning) and [ref2](https://github.com/openssl/openssl/issues/5042)). That would suggest the symlinking solution only nominally works, and allows MariaDB to start up. Perhaps it has been "working" because MariaDB, when launched from MariaDB4j, does not ever need to call into the openssl binary - MariaDB4j is most commonly used for integration tests, in which case, there's no point to using TLS/SSL for a localhost connection.

If MariaDB4j is not the one causing you to use multiple versions of openssl, then you will need something to manage these multiple openssl versions, regardless of whether you use MariaDB4j. In this regard, symlinking to `/usr/.../libssl.1.0.0.dylib` is only an afterthought. Only programs which need openssl 1.0.0 will look for it in `/usr/.../libssl.1.0.0.dylib`, so you cannot cause any harm by creating the symlink: to be precise, you cannot break anything which was previously working. <sub>Just to be clear in case you are not familiar with symlinks: the second argument is the target file which is created, meaning `ln -s /usr/lib/libssl.dylib /usr/.../libssl.1.0.0.dylib` does not affect the first file but it does create the second file.</sub>

If MariaDB4j is the one causing you to use multiple versions of openssl, then you are in the same boat as I and all the other users in #48 . This is the bottom line, and it is the big question. Is symlinking openssl 1.1.x to openssl 1.0.0 a safe maneuver? Practical experience suggests it does not matter in the case of MariaDB4j. Considering that the most common use of MariaDB4j, integration tests, rarely need to use SSL, it may be that no issues are encountered due to this fact. It would be useful to have someone who simultaneously uses MariaDB4j, SSL with their MariaDB connection, and Mac OS describe their experience.
